### PR TITLE
Add IpcStrategyWithBackup

### PR DESCRIPTION
**Pull request by rpdome**
Second part of https://microsoft-my.sharepoint-df.com/:w:/p/rapong/EUJd76fanC9Hgpic0UbLjvUBsFRv6jKt_uTz888YNUj07w?e=cJHbNz

This will allow us to try other ipc strategies if the main one fails (backup)

the backup will be best effort.
if they fail, it'll be like they're never there in the first place.

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    private val backup: List<IIpcStrategy>): IIpcStrategy{
+
+    override fun communicateToBroker(bundle: BrokerOperationBundle): Bundle? {
+        return try {
+            primary.communicateToBroker(bundle)
+        } catch (t: Throwable) {
+            for(ipc in backup) {
+                try {
+                    val result = ipc.communicateToBroker(bundle)
+                    SpanExtension.current().setAttribute(
```
Comment (rpdome) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt:50]: @shahzaibj I'm planning to query # of success flows that use primary vs backup

Does what i'm adding here make sense?

File: common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                        AttributeName.backup_ipc_used.name,
+                        ipc.type.name
+                    )
+                    return result
+                } catch (t: Throwable) {
+                    // do nothing.
```
Comment (fadidurah) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt]: Should we log individual exceptions from backups?

**Review by fadidurah — APPROVED** [review]
Minor suggestion

File: common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import android.os.Bundle
+import com.microsoft.identity.common.java.logging.Logger
+import com.microsoft.identity.common.java.opentelemetry.AttributeName
+import com.microsoft.identity.common.java.opentelemetry.SpanExtension
+
+/**
+ * An IPC Strategy wrapper with backups.
+ * If the primary one fails, the backups will be used "best-effort".
+ * If all mechanisms fail, only the exception of the main one will be thrown.
+ * */
```
Comment (somalaya) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt:34]:
```suggestion
 */
```

File: common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+class IpcStrategyWithBackup (
+    private val primary: IIpcStrategy,
```
Comment (somalaya) [inline common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/IpcStrategyWithBackup.kt:36]: nit : can we change the var names to primaryIpcStrategy, backupIpcStrategy ?

**Review by somalaya — APPROVED** [review]
LGTM. Just left some minor comments.

**Pull request merged**